### ISSUE TITLE
If two or more controllers are connected they are considered the same controller

### ISSUE THREAD

**Issue by Dukemz** [quinton-ashley/p5play#327]
![image](https://github.com/user-attachments/assets/6ba2d6b0-e374-475e-845b-f631f1715dbd)

If two controllers/gamepads are connected, p5play treats them as the same (in the attached screenshot, they are both considered to be contros[0]).

![image](https://github.com/user-attachments/assets/efe32e86-9073-49e5-8d81-61dd8be5a942)

I'm not completely sure why this happens? Possibly because gamepad index isn't taken into account?

**Comment by quinton-ashley** [comment]
What version of p5play are you using?

**Comment by Dukemz** [comment]
3.22.11, forgot to include in the post my bad

**Comment by quinton-ashley** [comment]
Thanks for reporting the issue but I don't know why it occurs.

Take a look at the `_Contros._onConnect` function:

https://github.com/quinton-ashley/p5play/blob/e587c7f9e78b50ed14046c5c15667f6bd3c1f68a/p5play.js#L10563

The indexes of controllers in p5play's `contros` array differ from the HTML5 `navigator.gamepads` array because I wanted reconnected controllers to be at the same index in the `contros` array and new controllers should be added to the next available slot in the controllers array (if previous controllers were removed).

If `contros[0]` already exists it shouldn't be possible to overwrite it. Can you try to set breakpoints in p5play.js to see why index is still 0 when the second controller is added?

**Comment by Dukemz** [comment]
The second time the _onConnect function is called it seems that `this` is an array with just 1 object
![image](https://github.com/user-attachments/assets/6e084aa3-8c98-4b6e-ad9e-ede12ca2d44e)
which as a result causes index to be 0.
Looking at the MDZN documentation, it says that controllers that are disconnected will be replaced in the array returned by `navigator.getGamepads()` with null, but will remain in the same index.
![image](https://github.com/user-attachments/assets/6972c53c-ef36-4225-bc67-312bba2d5149)
I'm still not entirely sure what's going on but I'll reply again when I've figured out some more details.

**Comment by Dukemz** [comment]
![image](https://github.com/user-attachments/assets/eb556ac9-ef4e-463c-9049-34a10bf1bc87)
This seems to be the most reliable fix. The index _should_ stay the same between disconnects/reconnects.

**Comment by Dukemz** [comment]
Bug still present in 3.23. Putting what Quinton said on the p5play discord server here for anyone not in the server.
![image](https://github.com/user-attachments/assets/e391e094-8d98-41c3-a410-8a94821a6e8f)

**Comment by quinton-ashley** [comment]
@Dukemz I will take another look at this soon.

**Comment by quinton-ashley** [comment]
@duckemz I think I fixed the controller error in v3.25.3.

Before, the condition of the for loop was `i < this.length`, which is wrong because the index value we want when controllers are initially connected is always `this.length`. For example if a controller is present at index 0, we want the second controller to be added at index 1.
```js
let index = 0;
for (let i = 0; i <= this.length; i++) {
    if (!this[i]) {
        index = i;
        break;
    }
}
this[index] = c;
log('contros[' + index + '] connected: ' + gp.id);
```
